src/mantine-core/src/PinInput/PinInput.tsx

```diff
@@ -190,16 +190,20 @@ export const PinInput = forwardRef<HTMLDivElement, PinInputProps>((props, ref) =
 
   const handleChange = (event: React.ChangeEvent<HTMLInputElement>, index: number) => {
     const inputValue = event.target.value;
-    const nextChar =
-      inputValue.length > 1 ? inputValue.split('')[inputValue.length - 1] : inputValue;
+    const nextCharOrValue =
+      inputValue.length === 2 ? inputValue.split('')[inputValue.length - 1] : inputValue;
```

Comment:
Would not it break the case when we paste values from regular keyboard, for example when `1234` is pasted?

Reply:
As the regular keyboard is fully handled by the `onPaste` event handler, logically it should not have any breaks. 
I also tried it out using `Ctrl + v` and `RightClick  + paste` that was ok too.